### ISSUE TITLE
Should we enforce a coding standard ?

### ISSUE THREAD

**Issue by alalazo** [LLNL/spack#337]
As the number of contributors to `spack` is growing I was wondering if it may be useful to enforce some coding standard on the project. In the long run it may help readability (and PR diffs will not be cluttered by style differences). What do you think?

**Comment by tgamblin** [comment]
I like the idea.  I've been hesitant to impose too much on contributors so far because, well, I want contributors :smile:.  But I think in the long run this will be good.  Ideally, the coding standard would be checkable on Travis so that the review burden is not too high.  Do you have a proposal for a particular standard/experience with any auto-checkers?

**Comment by alalazo** [comment]
I don't have much experience with python yet, so I may be missing some widely used tool. Taking a look around it seems that the two most used checkers are [pylint](http://www.pylint.org/) and [flake8](https://flake8.readthedocs.org/en/latest/). For the coding standard  they both default to PEP8 and every aspect can be overridden by a configuration file. It should be quite easy to integrate them with travis (e.g. `flake8 --count <directory>` sets the exit code to 1 if the total number of issues is not null).

**Comment by alalazo** [comment]
@tgamblin : this afternoon I had time to play a bit with various tools for code checking in python. In case you want to experiment with the configuration files for `pylint` and `flake8` the result is [in this branch](https://github.com/alalazo/spack/tree/qa/tox) (with some glitches still to be fixed for Travis).

###### Modifications to the test environment
I set up [tox](https://tox.readthedocs.org/en/latest/) for testing. On Travis I am running code coverage on the framework and submitting the results to [coveralls](https://coveralls.io/github/alalazo/spack) . `pylint` and `flake8` can be run locally e.g. `tox -e pylint`. As you will see they are "very talkative". I still have a random failure to be tracked on Travis, where `spack test concretize` happens to fail from time to time.

**Comment by alalazo** [comment]
@tgamblin : I have [here](https://github.com/alalazo/spack/tree/qa/pep8_only) a minimal change-set in case we just want to introduce `pep8` and coverage into the mix. To avoid breaking all the checks because of PEP8 non conformancy, I set up Travis to run a code style check only on python files that have been changed in the current PR. Generally, most part of the style issues (like in the compiler.py file that I used as a test) are trivial, and can be solved almost automatically using `autopep8`. Let me know if there are some other things you would like to try in the direction of an improved code checking.

**Comment by trws** [comment]
It wont help for checking, but for python code formatting I'm rather fond of [yapf](https://github.com/google/yapf), which applies the same approach of parsing to AST followed by formatted print-back as clang-format.

**Comment by alalazo** [comment]
@trws : I quite like the tool. I think it's a better option than `autopep8` or similar tools for correcting formatting issues.

**Comment by tgamblin** [comment]
@alalazo, @trws: Thanks for all the work on this. I think only checking changed files makes sense and the coverage stuff is really useful.  Is this ready for a PR?

One thing I've also thought about is something to audit packages in more depth -- Homebrew has something called [`brew audit`](https://github.com/Homebrew/homebrew/blob/master/share/doc/homebrew/Formula-Cookbook.md#audit-the-formula) that they use for this.  Maybe that could be added in to the tests later.

**Comment by alalazo** [comment]
@tgamblin : I 'll prepare a a PR from [here](https://github.com/alalazo/spack/tree/qa/pep8_only) if you are ok with it. I'll revert the notification e-mail in the process :smile:

**Comment by trws** [comment]
If we're doing it, some linting might not be a terrible idea.  Maybe just pull in pylint to enforce some stylistic but non-formatting rules?

**Comment by alalazo** [comment]
@trws : will do

**Comment by citibeth** [comment]
In terms of Spack-specific code standards... it is helpful to standardize the names of variants.  For example, "+doc" (or whatever) should build documentation for a packages, not "+doc" on one package, "+docs" on another and "+apidocs" on a third.

**Comment by tgamblin** [comment]
agree -- do you think the `doc`umentation is the right place for a best practices guide like this?  Want to start a PR with an extra top-level doc section on variant naming?

**Comment by trws** [comment]
I agree, but would, personally, make it a little stronger.  It should be both best practices for variant naming, and a list of globally valid variants that should *always* be used to enable, or disable, a given thing.  Actually, `doc` is a perfect example of one such, another might be `python`, for whether to build the python bindings in a given package or tree of packages, etc.

**Comment by tgamblin** [comment]
Seems like something for a top-level `.rst` in `lib/spack/docs`

**Comment by adamjstewart** [comment]
Most of the current users of Spack are also active developers. This may not always be the case. We may want to consider having a clearer delineation between "how to use Spack" and "how to develop Spack" in the documentation.

**Comment by adamjstewart** [comment]
Not to open a can of worms, but has there been any discussion on porting Spack to Python 3 in the future? Since we already require Python 2.6+, we may be able to get away with enforcing the use of some of the backported features. For example, if we require people to use
```python
'{0} {1}'.format('hello', 'world')
```
instead of
```python
'%s %s' % ('hello', 'world')
```
then that may save us from a lot of work when it comes time to transition. I would be willing to start a process like this on existing code.

**Comment by tgamblin** [comment]
I think that is a topic for a separate issue.  But FYI, `%s` is still in Python 3, so that's not the biggest portability concern.  The main thing will be to stop using `print` with no parens around the arguments.

Ideally I'd like Spack to work in both Python 2 and Python 3.

**Comment by adamjstewart** [comment]
Compatibility with Python 2 and 3 can be part of a coding standard. It's never too early to start!

But I agree that getting this transition started and setting up a test framework for it are well outside the bounds of this discussion. I'll save that for another day :)

**Comment by citibeth** [comment]
When I tried to use '{}'.format() in a PR, Travis CI complained that it was
not Python 2.6-compatible.  If/when we are willing to drop Python 2.6
support, I think we can move to Python3-compatible code in earnest.  Until
then, I'm not sure it's possible to be fully Python3-compliant and run on
Python 2.6.

I currently have access to MacOS El Capitan, CentOS 7 and Ubuntu 14.04LTS
--- all relatively new "long-term support" OS versions.  Of those three,
only Ubuntu 14.04 offers python3 out-of-the-box.

The purpose of Spack is to make it easy to install things.  Requiring a
Python3 installation to run Spack would work against that goal.  I think
it's simplest to consider Spack to be Python2-only for the forseeable
future.

(BTW.. I've already converted my application code to Python3, and I have
Spack install the Python3 needed for that code.  But I think Spack itself
needs to stay with Python2 for a good while longer).

-- Elizabeth

**Comment by tgamblin** [comment]
An *empty* format expression is not 2.6 compatible -- you have to put a number in there.  So `{0}` vs. `{}`.

I think we definitely need to stick with Python2 until it's gone (which might be never).  I was just saying that a lot of libs are moving to Python2 + 3 at the same time, per [this article](https://hynek.me/articles/python3-2016/) that was on hacker news recently.  Specifically, it claims (and I have not verified) that *"the vast majority of important libraries on PyPI are hybrid: they support both Python 2 and 3"*.  I like to shoot to be inline with peoples' exceptions, so that seems like the thing to aim for.

**Comment by tgamblin** [comment]
We have now both merged @alalazo's `flake8` checker and I've completely `PEP8`-ified Spack in #1496.  The `README.md` has also been updated to tell people what the coding standard is, and also thanks to @alalazo, Travis has separate checks for style and correctness now.  I think this can be closed.  

RE: Python 3 support, yes, I think that is the next big hurdle.  I would like to address that as @citibeth suggests _after_ we can drop Python 2.6 support.  AFAIK, there are still HPC systems that have 2.6 as the default python on the front-end nodes.  It's very sad.  I'd be interested in doing a survey of major HPC sites at some point to see what the breakdown is, so that we can decide to move on sooner rather than later.

Thanks everyone!

**Comment by adamjstewart** [comment]
> AFAIK, there are still HPC systems that have 2.6 as the default python on the front-end nodes. It's very sad.

Our _newest_ cluster runs CentOS 6 and comes with Python 2.6.6. Our _old_ cluster runs CentOS 5 and comes with Python 2.4.3. It's so old that `python --version` wasn't even a valid command.

This fall/winter we're replacing our old cluster with a new one running CentOS 7. Once development on that stabilizes, we'll be updating the other cluster to CentOS 7 as well.

**Comment by citibeth** [comment]
I'm saying this as someone who has converted to 100% Python3 for all my
scientific work...

I'm not sure if now is the right time to move Spack to Python3.  Let's
see...

  *) RHEL 7: No Python3.  (Available only through the non-standard EPEL
repo, don't count on sysadmins to have installed it).
  *) Ubuntu 16.04: Python3 is the default, but both python and python3 are
available in the normal repo.
  *) openSUSE: Python3 is available, but not the default.  They have a nice
chart of Python3 compatibility:
    https://en.opensuse.org/openSUSE:Python_3_Status
 *) OS X Yosemite: You must install Python3 with brew, macports, anaconda,
etc.

Python 2.7 is still universal, if you want a script to "just work" on these
systems.  That is why I build "system" software (like Spack) in Python2,
while building scientific analysis in Python3.

I would recommend that we begin moving to Python 2/3 compatible code, even
if we're only running it on Python2 for the time being.  For example:
    http://python-future.org/compatible_idioms.html
..and we can also include `six` in Spack, that will ease the transition.
The more compatible our code is now, the less pain there will be in the
future.

I don't think we should be requiring contributions to run in Python3 at
this time.  That will increase the testing burden on contributors.

-- Elizabeth

**Comment by adamjstewart** [comment]
I agree, Spack will be Python 2.7/3 compatible long before it is 3 only. I think we still need to wait a bit before we drop 2.6 compatibility though. In the meantime, I've been using things like `'--prefix={0}'.format(prefix)` instead of `'--prefix=%s' % prefix` to ease the transition down the road.

**Comment by tgamblin** [comment]
@citibeth: I agree.  We wouldn't move to Python 3 exclusively until 2.7 is long gone.  We would maintain 2.7 compatibility and go 2.7/3 like @adamjstewart says.  Sorry if I implied otherwise.

Also agree on Python 3 compatibility for contributions.  We should wait _at least_ until after dropping 2.6 support before thinking about it.

**Comment by citibeth** [comment]
>
> I've been using things like '--prefix={0}'.format(prefix) instead of '--prefix=%s'
> % prefix to ease the transition down the road


The use of '%s' has not been deprecated, and there are currently no plans
to do so.  This is a separate issue from Python2 vs. 3.

**Comment by adamjstewart** [comment]
Huh, I don't know why I thought it was deprecated. I know print is a function now instead of a statement. I wonder how much we'll actually need to change to add Python 3 support.

**Comment by citibeth** [comment]
The biggest thing we'll have to change is the print statement.  We can do
that today by adding `from __future__ import print_function` to the top of
every `.py` file.

**Comment by trws** [comment]
Unfortunately our center is one such until TOSS3 drops.  It's really quite disturbing, in that 2.6 can't be considered reasonably securable anymore, but we are where we are I suppose.  If it has to be done before 2.6 can be dropped, six is an option, but man would it be better to just get to 2.7.
